# dorinemerlat/exogap: Citations

## [EXOGAP](https://github.com/dorinemerlat/exogap)

  > Merlat, D. (2023) EXOGAP. doi: [10.5281/zenodo.8407927](https://doi.org/10.5281/zenodo.8407927).
## [Nextflow](https://www.nextflow.io/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: [10.1038/nbt.3820](https://www.nature.com/articles/nbt.3820). PubMed PMID: [28398311](https://pubmed.ncbi.nlm.nih.gov/28398311/).

## [nf-core](https://nf-co.re/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: [10.1038/s41587-020-0439-x](https://www.nature.com/articles/s41587-020-0439-x). PubMed PMID: [32055031](https://pubmed.ncbi.nlm.nih.gov/32055031/).

## Pipeline tools

- [BiGEst APIs](https://lbgi.fr/api/index.rvt)

  > Kress, A. (2023). BiGEst APIs. [https://lbgi.fr/api/index.rvt](https://lbgi.fr/api/index.rvt)

- [Bioawk](https://github.com/lh3/bioawk)

  > Li, H. (2015). Bioawk. github: [lh3/bioawk](https://github.com/lh3/bioawk)

- [BUSCO](https://busco.ezlab.org/)

  > Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update : Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647‑4654. doi: [10.1093/molbev/msab199](https://doi.org/10.1093/molbev/msab199)

- [Jupyter Notebook](https://jupyter.org/)

  > Kluyver, T., Ragan-Kelley, B., Pérez, F., Granger, B., Bussonnier, M., Frederic, J., Kelley, K., Hamrick, J., Grout, J., Corlay, S., Ivanov, P., Avila, D., Abdalla, S., & Willing, C. (2016). Jupyter Notebooks – a publishing format for reproducible computational workflows. In F. Loizides & B. Schmidt (Éds.), Positioning and Power in Academic Publishing : Players, Agents and Agendas (p. 87‑90). IOS Press.

- [Python](https://www.python.org/)

  > Van Rossum, G., & Drake, F. L. (2009). Python 3 Reference Manual. CreateSpace.

- [RepeatMasker](http://www.repeatmasker.org)

  > Smit, A. F. A., Hubley, R., & Green, P. (2015). RepeatMasker Open-4.0. [http://www.repeatmasker.org](http://www.repeatmasker.org)

- [RepeatModeler](https://www.repeatmasker.org/RepeatModeler/)

  > Flynn, J. M., Hubley, R., Goubert, C., Rosen, J., Clark, A. G., Feschotte, C., & Smit, A. F. (2020). RepeatModeler2 for automated genomic discovery of transposable element families. Proceedings of the National Academy of Sciences, 117(17), 9451‑9457. doi: [10.1073/pnas.1921046117](https://doi.org/10.1073/pnas.1921046117)

## Python packages

- [Biopython](https://biopython.org/)

  > Cock, P. J. A., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B., & De Hoon, M. J. L. (2009). Biopython : Freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics, 25(11), 1422‑1423. doi: [10.1093/bioinformatics/btp163](https://doi.org/10.1093/bioinformatics/btp163)

- [Environment for Tree Exploration 3 (ETE3)](http://etetoolkit.org/)

  > Huerta-Cepas, J., Serra, F., & Bork, P. (2016). ETE 3 : Reconstruction, Analysis, and Visualization of Phylogenomic Data. Molecular Biology and Evolution, 33(6), 1635‑1638. doi: [10.1093/molbev/msw046](https://doi.org/10.1093/molbev/msw046)

- [Kaleido](https://github.com/plotly/Kaleido)

  > Mease, J. (2021). Kaleido (v0.2.1.post1). github: [plotly/Kaleido](https://github.com/plotly/Kaleido)

- [Pandas](https://pandas.pydata.org/)

> Team, T. P. D. (2023). pandas-dev/pandas : Pandas (v2.1.1). Zenodo. doi: [10.5281/ZENODO.8364959](https://doi.org/10.5281/ZENODO.8364959)

- [Plotly](https://plot.ly)

  > Inc, P. T. (2015). Collaborative data science [Logiciel]. Plotly Technologies Inc. [https://plot.ly](https://plot.ly)

- [Py Fasta Validator](https://github.com/linsalrob/py_fasta_validator)

  > Edwards, R. (2019). linsalrob/fasta_validator : Initial Release (v0.1). Zenodo. doi: [10.5281/ZENODO.2532044](https://doi.org/10.5281/ZENODO.2532044)

- [Python-newick](https://pypi.org/project/newick/)

  > Forkel, R. (2023). Python-newick (1.9.0). github: [dlce-eva/python-newick](https://github.com/dlce-eva/python-newick)

- [Request](https://github.com/psf/requests)

> Reitz, K. (2023). Requests (2.31.0). github: [psf/requests](https://github.com/psf/requests)

## Database

- [RepBase](https://www.girinst.org/repbase/)

  > Bao, W., Kojima, K. K., & Kohany, O. (2015). Repbase Update, a database of repetitive elements in eukaryotic genomes. Mobile DNA, 6(1), 11. doi:[10.1186/s13100-015-0041-9](https://doi.org/10.1186/s13100-015-0041-9)

- [Dfam](https://www.dfam.org/home)

> Storer, J., Hubley, R., Rosen, J., Wheeler, T. J., & Smit, A. F. (2021). The Dfam community resource of transposable element families, sequence models, and genome annotations. Mobile DNA, 12(1). doi: [10.1186/S13100-020-00230-Y](https://doi.org/10.1186/S13100-020-00230-Y)

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://anaconda.org/bioconda)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: [10.1038/s41592-018-0046-7](https://www.nature.com/articles/s41592-018-0046-7). PubMed PMID: [29967506](https://pubmed.ncbi.nlm.nih.gov/29967506/).

- [BioContainers](https://biocontainers.pro/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: [28379341](https://pubmed.ncbi.nlm.nih.gov/28379341/); PubMed Central PMCID: [PMC5870671](https://pubmed.ncbi.nlm.nih.gov/28379341/).

- [Docker](https://www.docker.com/)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: [10.5555/2600239.2600241](https://dl.acm.org/doi/10.5555/2600239.2600241).

- [Singularity](https://docs.sylabs.io/guides/3.5/user-guide/introduction.html)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: [10.1371/journal.pone.0177459](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0177459). eCollection 2017. PubMed PMID: [28494014](https://pubmed.ncbi.nlm.nih.gov/28494014/); PubMed Central PMCID: [PMC5426675](https://pubmed.ncbi.nlm.nih.gov/28494014/).
