![nf-core/exogap](./docs/images/exogap_logo_light.png#gh-light-mode-only)
![nf-core/exogap](./docs/images/exogap_logo_dark.png#gh-dark-mode-only)

---

## Introduction

**EXOGAP** (EXotic Organism Genome Annotation Pipeline) is a Nextflow workflow designed for the comprehensive annotation of genomes from non-model species. It integrates major tools for repetitive element identification, protein-coding gene and non-coding gene prediction. EXOGAP manages  data conversion and communication between all these tools, filters and selects the most reliable results, and generates insightful statistics and plots to facilitate easy analysis of annotation quality. It allows annotation of multiple genomes from a group of closely related species in a single run, while adapting the data to used for each genome and using the predictions of other genomes to enhance prediction quality of each genome. Moreover, It also ensures that all annotations are formatted for easy deposition in public databanks. As a Nextflow workflow, EXOGAP efficiently utilizes Docker or Singularity to install and configure all the necessary programs, simplifying the setup process for researchers.

This is the major steps of EXOGAP:
1. *Data Preparation*: Start by preparing the necessary data, calculating genome statistics, and obtaining taxonomic information.
2. *Repetitive Element Annotation*: annotate repetitive elements within the genome using RepeatModeler and RepeatMasker.
3. *Protein-Coding Gene Annotation*: annotate protein-coding genes using MAKER2, AUGUSTUS and SNAP, Blast and InterProScan (optionnaly: Blast2GO).
4. *Non-Coding Gene Annotation*: annotate non-coding genes with tRNAscan-SE, RNAmmer, Barrnap, Infernal, SnoScan.
5. *Annotation Formatting*: format the annotations and sequence files to meet the requirements of major public databanks.
6. *Statistical Analysis*: generate informative statistics and analyses to assess the quality of genome annotation.

<!-- TODO: Add a metro map of EXOGAP -->
## Installation

1. To use EXOGAP, you must ensure that you have a functional installation of[Nextflow](https://www.nextflow.io/) (version >= 23.09.2-edge.5883) and [Singularity](https://sylabs.io/singularity/) (version >= 3.0).

2. Once Nextflow is installed, you need to update it to use a *edge* release. Indeed, EXOGAP uses some features that are not yet available in the *stable* release of Nextflow. To do so, run the following command:
```bash
export NXF_EDGE=1
nextflow self-update
```
Don't worry, the *edge* release works very well with EXOGAP and will not cause any issues.

3. Clone the GitHub repository of EXOGAP:
```bash
https://github.com/dorinemerlat/exogap.git
```

## Usage

1. Begin by creating a samplesheet for the genome assemblies you wish to annotate. This samplesheet should be in a comma-separated format with three columns, where each row corresponds to a genome file to annotate. This is an example of `samplesheet.csv`.

  ```csv
  name,taxid,fasta
  Polydesmus complanatus,510027,polydesmus-complanatus.fa
  ```

  | **Column** | **Description** |
  |------------|-----------------|
  | `name`     | Genome to use. This is not necessarily the official name of the specie, but will be used in graphs and tables generated by the pipeline (alphnanumeric characters, spaces, - and _ accepted).  |
  | `taxid`    | It must be valid and available in the NCBI taxonomic database. If no taxid is available for the species or strain, then enter the taxid of the nearest parent taxon. The same taxid can be entered in several lines, if several genomes to be annotated share the same taxon.      |
  | `fasta`    | Name of the fasta file associated with the genome. By default, EXOGAP looks for this file in the `/data/genomes/` folder. If you wish to use a different location, please specify the full path of the file.        |

  The order of the columns in the samplesheet doesn't matter, just update the header to match the order you prefer. In addition, it is mandatory to fill in all 3 columns for each genome. As for the rows, the file must at least contain the header and the information for one genome.

2. Then, provide pipeline parameters via the config file `nextflow.config`.

3. Now, you can run the pipeline using:

  ```bash
  nextflow exogap \
    -profile <docker/singularity/.../institute> \
    --input samplesheet.csv \
    --config nextflow.config
    --outdir <OUTDIR>
  ```

## Pipeline output

Coming soon.

## Credits

EXOGAP is written and maintained by Dorine MERLAT (dorine.merlat@etu.unistra.fr).

I thank the following people for their extensive assistance in the development of EXOGAP: Odile Lecompte and Arnaud Kress.

## Citations

<!-- TODO nf-core: Add citation for pipeline after first release. Uncomment lines below and update Zenodo doi and badge at the top of this file. -->
<!-- If you use  nf-core/exogap for your analysis, please cite it using the following doi: [10.5281/zenodo.XXXXXX](https://doi.org/10.5281/zenodo.XXXXXX) -->

<!-- TODO nf-core: Add bibliography of tools and data used in your pipeline -->

An extensive list of references for the tools used by the pipeline can be found in the [`CITATIONS.md`](CITATIONS.md) file.
