![nf-core/exogap](./docs/images/exogap_logo_light.png#gh-light-mode-only)
![nf-core/exogap](./docs/images/exogap_logo_dark.png#gh-dark-mode-only)

---

## Introduction

**EXOGAP** is a bioinformatics pipeline that annotate one or several genomes of non-model species. It combines several bioinformatics tools to annotate the repetitive elements, the protein-coding genes and the non-coding genes. It also generates some statistics and analysis about genome annotation. It based on Nextflow to make the pipeline portable and reproducible. It comes with docker/singularity containers making installation trivial and results highly reproducible.

1. Prepare data, calculate some statistics about genomes and get taxonomic informations.
2. Annotate the repetitive elements.
3. Annotate the protein-coding genes.
4. Annotate the non-coding genes.
5. Formate the annotations and sequence files in the format of main public databank.
6. Generate some statistics and analysis about genome annotation.

<!-- TODO: Add a metro map of EXOGAP -->
## Usage

To use EXOGAP, it is necessary to have a working installation of [Nextflow](https://www.nextflow.io/) (version >= 20.01.0), [Docker](https://www.docker.com/) or [Singularity](https://sylabs.io/singularity/) (version >= 3.0).

The edge version of Nextflow is needed to use the pipeline. To update it, run the following command:

```bash
export NXF_EDGE=1
nextflow self-update
```


First, prepare a samplesheet information about the genome assemblies you would like to annotate before running the pipeline. It has to be a comma-separated file with 3 columns, where each row represents a genome file to annotate.

`samplesheet.csv`:

```csv
name,taxid,fasta
Polydesmus complanatus,510027,polydesmus-complanatus.fa
```
 | **Column** | **Description** |
|------------|-----------------|
| `name`     | Genome to use. This is not necessarily the official name of the specie, but will be used in graphs and tables generated by the pipeline (alphnanumeric characters, spaces, - and _ accepted).  |
| `taxid`    | It must be valid and available in the NCBI taxonomic database. If no taxid is available for the species or strain, then enter the taxid of the nearest parent taxon. The same taxid can be entered in several lines, if several genomes to be annotated share the same taxon.      |
| `fasta`    | Name of the fasta file associated with the genome. By default, EXOGAP looks for this file in the `/data/genomes/` folder. If you wish to use a different location, please specify the full path of the file.        |

Then, provide pipeline parameters via the config file `nextflow.config`. It allows to indicate the computing resources to use and to modularize the pipeline by giving additional databases, paying programs...

Now, you can run the pipeline using:

```bash
nextflow exogap \
   -profile <docker/singularity/.../institute> \
   --input samplesheet.csv \
   --config nextflow.config
   --outdir <OUTDIR>
```

## Pipeline output

To see the results of an example test run with a full size dataset refer to the [results](https://nf-co.re/exogap/results) tab on the nf-core website pipeline page.
For more details about the output files and reports, please refer to the
[output documentation](https://nf-co.re/exogap/output).

## Credits

EXOGAP was originally written by Dorine MERLAT.

We thank the following people for their extensive assistance in the development of this pipeline: Odile Lecompte and Arnaud Kress.

## Citations

<!-- TODO nf-core: Add citation for pipeline after first release. Uncomment lines below and update Zenodo doi and badge at the top of this file. -->
<!-- If you use  nf-core/exogap for your analysis, please cite it using the following doi: [10.5281/zenodo.XXXXXX](https://doi.org/10.5281/zenodo.XXXXXX) -->

<!-- TODO nf-core: Add bibliography of tools and data used in your pipeline -->

An extensive list of references for the tools used by the pipeline can be found in the [`CITATIONS.md`](CITATIONS.md) file.
